daemon/guestos_mgr.c

```diff
@@ -222,40 +224,60 @@ guestos_mgr_delete(guestos_t *os)
 /******************************************************************************/
 
 static void
-download_complete_cb(bool complete, unsigned int count, guestos_t *os, UNUSED void *data)
+download_complete_cb(bool complete, unsigned int count, guestos_t *os, void *data)
 {
-	IF_NULL_RETURN_ERROR(os);
-	IF_TRUE_RETURN_TRACE(count > 0);
+	int *resp_fd = data;
+	ASSERT(resp_fd);
+
+	control_message_t resp = CONTROL_RESPONSE_GUESTOS_MGR_INSTALL_FAILED;
+
+	IF_NULL_GOTO_ERROR(os, out);
 
-	if (complete) {
-		if (guestos_images_flash(os) < 0)
+	if (complete && count > 0) {
+		if (guestos_images_flash(os) < 0) {
 			WARN("%s %s", GUESTOS_MGR_UPDATE_TITLE, GUESTOS_MGR_UPDATE_FLASH_FAILED);
-		else
+		} else {
 			INFO("%s %s", GUESTOS_MGR_UPDATE_TITLE, GUESTOS_MGR_UPDATE_SUCCESS);
+			resp = CONTROL_RESPONSE_GUESTOS_MGR_INSTALL_COMPLETED;
+		}
 	} else {
 		WARN("%s %s", GUESTOS_MGR_UPDATE_TITLE, GUESTOS_MGR_UPDATE_FAILED);
 	}
+out:
+	if (control_send_message(resp, *resp_fd) < 0)
+		WARN("Could not send response to fd=%d", *resp_fd);
+	mem_free(resp_fd);
```

Comment:
freeing stack-allocated pointer?

Reply:
when cb is done it has to free its associated data yes.